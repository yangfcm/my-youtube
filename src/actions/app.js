import { defAxios as axios, maxResults } from "../settings";
import {
  FETCH_CHANNEL,
  FETCH_CHANNEL_SUBSCRIPTION,
  FETCH_PLAY_LIST,
  FETCH_VIDEOS,
  SEARCH_VIDEOS,
  FETCH_CHANNEL_INTRO,
  FETCH_PLAY_LIST_DETAIL,
  FETCH_VIDEO,
  FETCH_COMMENTS,
  FETCH_COMMENTS_DISABLED,
  CATCH_ERROR,
  CLEAR_ERROR,
  SUBSCRIBE_CHANNEL,
  UNSUBSCRIBE_CHANNEL,
  FETCH_REPLIES,
} from "./types";

export const searchVideos = (filter, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/search", {
        params: {
          ...filter,
          part: "snippet",
          key: process.env.REACT_APP_API_KEY,
          maxResults: 15,
          pageToken,
        },
      });
      dispatch({
        type: SEARCH_VIDEOS,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch videos list",
      });
    }
  };
};

export const fetchVideos = (filter, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...filter,
          part: "snippet,statistics,contentDetails",
          key: process.env.REACT_APP_API_KEY,
          maxResults: 15,
          pageToken,
        },
      });
      dispatch({
        type: FETCH_VIDEOS,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch videos list",
      });
    }
  };
};

/** Fetch comments by video id */
export const fetchComments = (videoId, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/commentThreads", {
        params: {
          part: "snippet",
          key: process.env.REACT_APP_API_KEY,
          videoId,
          maxResults,
          pageToken,
          order: "relevance",
        },
      });
      dispatch({
        type: FETCH_COMMENTS,
        payload: response.data,
      });
    } catch (e) {
      const errorReason = e.response.data.error.errors[0].reason;
      if (errorReason === "commentsDisabled") {
        // Consider comment is disabled by publisher.
        dispatch({
          type: FETCH_COMMENTS_DISABLED,
          payload: errorReason,
        });
      } else {
        dispatch({
          type: CATCH_ERROR,
          payload: "Failed to fetch comments",
        });
      }
    }
  };
};

/** Fetch replies to a comment */
export const fetchCommentReplies = (commentId, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/comments", {
        params: {
          part: "snippet",
          key: process.env.REACT_APP_API_KEY,
          parentId: commentId,
          maxResults,
          pageToken,
        },
      });
      dispatch({
        type: FETCH_REPLIES,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch replies",
      });
    }
  };
};

/** Fetch a vidoe by id */
export const fetchVideo = (videoId) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          part: "snippet,statistics",
          key: process.env.REACT_APP_API_KEY,
          id: videoId,
        },
      });
      dispatch({
        type: FETCH_VIDEO,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch video",
      });
    }
  };
};

/** Fetch playlist by ChannelId, if channel id is not provided, fetch the playlist of the log-in user*/
export const fetchPlaylist = (pageToken, channelId = "") => {
  return async (dispatch) => {
    try {
      let response;
      if (channelId) {
        response = await axios.get("/playlists", {
          params: {
            part: "snippet,contentDetails,status",
            maxResults,
            channelId,
            pageToken,
            key: process.env.REACT_APP_API_KEY,
          },
        });
      } else {
        const accessToken = localStorage.getItem("access_token");
        response = await axios.get("/playlists", {
          headers: {
            Authorization: accessToken,
          },
          params: {
            part: "snippet,contentDetails,status",
            maxResults,
            pageToken,
            key: process.env.REACT_APP_API_KEY,
            mine: true,
          },
        });
      }
      dispatch({
        type: FETCH_PLAY_LIST,
        payload: response.data,
      });
    } catch (e) {
      if (e.response.data.error && e.response.data.error.code === 404) {
        dispatch({
          type: CATCH_ERROR,
          payload: "No playlist in this channel",
        });
        return;
      }
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch playlist",
      });
    }
  };
};

/** Fetch play list detail */
export const fetchPlaylistDetail = (playlistId, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/playlistItems", {
        params: {
          part: "snippet,contentDetails,status",
          maxResults: 8,
          key: process.env.REACT_APP_API_KEY,
          playlistId,
          pageToken,
        },
      });
      dispatch({
        type: FETCH_PLAY_LIST_DETAIL,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch playlist",
      });
    }
  };
};

/** Fetch subscribed channels by the current authorized user */
export const fetchChannel = (pageToken) => {
  const accessToken = localStorage.getItem("access_token");

  return async (dispatch) => {
    try {
      const response = await axios.get("/subscriptions", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet",
          maxResults: 50,
          // channelId: process.env.REACT_APP_MY_CHANNEL_ID,
          key: process.env.REACT_APP_API_KEY,
          pageToken,
          order: "alphabetical",
          mine: true,
        },
      });
      dispatch({
        type: FETCH_CHANNEL,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch channels",
      });
    }
  };
};

/** Given a channel id if it is subscribed by the user with access token, return the channel. Otherwise, return null */
export const fetchChannelSubscription = (channelId, accessToken) => {
  // const accessToken = localStorage.getItem("access_token");
  // if (!accessToken) {
  //   return {
  //     type: FETCH_CHANNEL_SUBSCRIPTION,
  //     payload: null,
  //   };
  // }

  return async (dispatch) => {
    try {
      const response = await axios.get("/subscriptions#list", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet",
          forChannelId: channelId,
          key: process.env.REACT_APP_API_KEY,
          mine: true,
        },
      });
      // console.log(response.data);
      if (response.data.items.length === 0) {
        // If the channel is not subscribed by the current authorized user
        dispatch({
          type: FETCH_CHANNEL_SUBSCRIPTION,
          payload: false,
        });
      } else {
        // If the channel is not subscribed by the current authorized user

        dispatch({
          type: FETCH_CHANNEL_SUBSCRIPTION,
          payload: true,
        });
      }
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch channel's subscription",
      });
    }
  };
};

/** Fetch channel intro */
export const fetchChannelIntro = (channelId) => {
  return async (dispatch) => {
    try {
      const response = await axios.get(`/channels`, {
        params: {
          part: "snippet,statistics",
          id: channelId,
          key: process.env.REACT_APP_API_KEY,
        },
      });
      dispatch({
        type: FETCH_CHANNEL_INTRO,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to fetch channel's intro",
      });
    }
  };
};

/** Subscribe a channel */
export const subscribeChannel = (channelId, accessToken) => {
  return async (dispatch) => {
    const response = await axios.post(
      "/subscriptions",
      {
        snippet: {
          resourceId: {
            kind: "youtube#channel",
            channelId: channelId,
          },
        },
      },
      {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet",
        },
      }
    );
    dispatch({
      type: SUBSCRIBE_CHANNEL,
      payload: response.data,
    });
    try {
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to subscribe channel",
      });
    }
  };
};

export const unsubscribeChannel = (channelId, accessToken) => {
  return async (dispatch) => {
    try {
      /** Use channelId to get subscription id */
      const subscriptionRes = await axios.get("/subscriptions#list", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet",
          forChannelId: channelId,
          key: process.env.REACT_APP_API_KEY,
          mine: true,
        },
      });

      if (subscriptionRes.data.items.length === 0) {
        throw Error("Channel not subscribed");
      }
      const subscriptionId = subscriptionRes.data.items[0].id;

      await axios.delete("/subscriptions", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          id: subscriptionId,
        },
      });
      dispatch({
        type: UNSUBSCRIBE_CHANNEL,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Failed to unsubscribe channel",
      });
    }
  };
};

/** Clear Error */
export const clearError = () => {
  return (dispatch) => {
    dispatch({
      type: CLEAR_ERROR,
    });
  };
};
