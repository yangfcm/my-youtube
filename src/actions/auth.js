import { SIGN_IN, SIGN_OUT, CATCH_ERROR } from "./types";
import { defAxios as axios } from "../settings";

export const signIn = (user) => {
  return async (dispatch) => {
    // console.log("action", user);
    try {
      if (!user) {
        throw Error("Authorization failed");
      }
      const accessToken = `${user.xc.token_type} ${user.xc.access_token}`;
      localStorage.setItem("access_token", accessToken);
      const response = await axios.get(
        "https://www.googleapis.com/oauth2/v1/userinfo?alt=json",
        {
          headers: {
            Authorization: accessToken,
          },
        }
      );
      dispatch({
        type: SIGN_IN,
        payload: {
          email: response.data.email,
          username: response.data.name,
          lastName: response.data.family_name,
          firstName: response.data.given_name,
          avatar: response.data.picture,
        },
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Authorization failed",
      });
    }
  };
};

export const signOut = () => {
  localStorage.removeItem("access_token");
  return {
    type: SIGN_OUT,
  };
};
