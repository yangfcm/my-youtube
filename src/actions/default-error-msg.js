export const DEFAULT_ERROR_MSG = {
  FAILED_TO_FETCH_VIDEO: "Failed to fetch video",
  FAILED_TO_SEARCH: "Failed to fetch resources",
  FAILED_TO_FETCH_COMMENTS: "Failed to fetch comments",
  FAILED_TO_FETCH_CHANNEL: "Failed to fetch channel",
  FAILED_TO_FETCH_CHANNEL_SUBSCRIPTION:
    "Failed to fetch subscription status for the channel",
  FAILED_TO_FETCH_PLAYLIST: "Failed to fetch play list",
  FAILED_TO_FETCH_PLAYLIST_DETAIL: "Failed to fetch videos in the play list",
  FAILED_TO_SERCH: "Failed to search resources",
  FAILED_TO_ADD_COMMENT: "Publishing comment is failed",
  FAILED_TO_FETCH_REPLIES: "Failed to fetch replies",
  FAILED_TO_SUBSCRIBE: "Failed to subscribe the channel",
  FAILED_TO_UNSUBSCRIBE: "Failed to unsubscribe the channel",
  COMMENT_DISABLED: "Comment is disabled by the author",
};
