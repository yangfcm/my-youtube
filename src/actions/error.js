import { CLEAR_ERROR } from "./types";

/** Clear Error */
export const clearError = () => {
  return {
    type: CLEAR_ERROR,
  };
  // return (dispatch) => {
  //   dispatch({
  //     type: CLEAR_ERROR,
  //   });
  // };
};
