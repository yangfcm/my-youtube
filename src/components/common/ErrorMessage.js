import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const renderError = (error) => {
  return (
    <div className="text-danger text-center my-3 py-3">
      <FontAwesomeIcon icon="exclamation-triangle" className="mx-2" />
      {error.displayMessage}
    </div>
  );
};

const ErrorMessage = ({ error }) => {
  // console.log(process.env);
  if (process.env.NODE_ENV === "development") {
    console.log(error);
  }
  return renderError(error);
};

export default ErrorMessage;
