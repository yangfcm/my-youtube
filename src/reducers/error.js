import { CATCH_ERROR, CLEAR_ERROR } from "../actions/types";

export const errorReducer = (state = null, action) => {
  switch (action.type) {
    case CATCH_ERROR:
      return action.payload;
    case CLEAR_ERROR:
      return null;
    default:
      return state;
  }
};
