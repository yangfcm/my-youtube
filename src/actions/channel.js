import { defAxios as axios } from "../settings";
import {
  FETCH_CHANNEL,
  FETCH_CHANNEL_SUBSCRIPTION,
  FETCH_CHANNEL_INTRO,
  SUBSCRIBE_CHANNEL,
  UNSUBSCRIBE_CHANNEL,
  CATCH_ERROR,
} from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";
