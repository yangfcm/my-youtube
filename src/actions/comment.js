import { defAxios as axios, maxResults } from "../settings";
import {
  ADD_COMMENT,
  FETCH_COMMENTS,
  FETCH_REPLIES,
  FETCH_COMMENTS_DISABLED,
  CATCH_ERROR,
} from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";

/**
 * Fetch comments under a video
 * If accessToken is provided, the comments by authorized user will be displayed on the top.
 */
export const fetchComments = (
  videoId,
  pageToken,
  accessToken,
  order = "relevance"
) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/commentThreads", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet",
          videoId,
          maxResults,
          pageToken,
          order: order === "relevance" ? "relevance" : "time",
        },
      });
      dispatch({
        type: FETCH_COMMENTS,
        payload: response.data,
      });
    } catch (e) {
      if (!e.response) {
        dispatch({
          type: CATCH_ERROR,
          payload: DEFAULT_ERROR_MSG.FAILED_TO_FETCH_COMMENTS,
        });
        return;
      }
      if (e.response.data.error.errors[0].reason === "commentsDisabled") {
        // Comment is disabled.
        dispatch({
          type: FETCH_COMMENTS_DISABLED,
          payload: DEFAULT_ERROR_MSG.COMMENT_DISABLED,
        });
      } else {
        dispatch({
          type: CATCH_ERROR,
          payload: e.response.data.error,
        });
      }
    }
  };
};

/** Add a comment to video */
export const addComment = (channelId, videoId, commentText, accessToken) => {
  return async (dispatch) => {
    try {
      const requestBdody = {
        snippet: {
          channelId,
          videoId,
          topLevelComment: {
            snippet: {
              textOriginal: commentText,
            },
          },
        },
      };
      const response = await axios.post("/commentThreads", requestBody, {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet",
        },
      });
      dispatch({
        type: ADD_COMMENT,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: e.response.data.error,
      });
    }
  };
};

/** Fetch replies to a comment */
export const fetchCommentReplies = (commentId, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/comments", {
        params: {
          part: "snippet",
          parentId: commentId,
          maxResults,
          pageToken,
        },
      });
      dispatch({
        type: FETCH_REPLIES,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: e.response.data.error,
      });
    }
  };
};
