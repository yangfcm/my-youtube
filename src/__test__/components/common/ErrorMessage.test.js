import React from "react";
import { shallow } from "enzyme";
import ErrorMessage from "components/common/ErrorMessage";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { DEFAULT_ERROR_MSG } from "actions/default-error-msg";
import { error } from "__test__/fixtures/error";

describe("Test ErrorMessage component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <ErrorMessage
        error={{
          ...error.response.data.error,
          displayMessage: DEFAULT_ERROR_MSG.FAILED_TO_FETCH_VIDEO,
        }}
      />
    );
  });
  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should have exclamation icon on ErrorMessage component", () => {
    const fontComp = wrapper.find(FontAwesomeIcon);
    expect(fontComp.exists()).toBe(true);
    expect(fontComp.prop("icon")).toBe("exclamation-triangle");
  });

  it("should display correct error message", () => {
    const errorComp = wrapper.find(".text-danger");
    expect(errorComp.text()).toContain(DEFAULT_ERROR_MSG.FAILED_TO_FETCH_VIDEO);
  });
});
