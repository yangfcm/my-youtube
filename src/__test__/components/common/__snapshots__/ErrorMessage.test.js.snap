// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ErrorMessage component should match snapshot 1`] = `
<div
  className="text-danger text-center my-3 py-3"
>
  <FontAwesomeIcon
    border={false}
    className="mx-2"
    fixedWidth={false}
    flip={null}
    icon="exclamation-triangle"
    inverse={false}
    listItem={false}
    mask={null}
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    swapOpacity={false}
    symbol={false}
    title=""
    transform={null}
  />
  Failed to fetch video
</div>
`;
